### DISCUSSION TITLE
《LLAMMA - crvUSD 的新清算機制》討論串

### DISCUSSION THREAD

**Discussion by yurenju** [lun-dao/LunDAO#100] [General]
這是《LLAMMA - crvUSD 的新清算機制》的討論串，歡迎在這邊討論各種疑問或想法。

**Comment by ksin751119** [comment]
你好，謝謝你分享這篇詳細說明的文章，讓我受益匪淺！
其中有一個問題一直想不明白，在程式碼中也找不太到相關的訊息。

假設 AMM Pool 的交易對是 crvUSD <-> ETH。當在對 AMM 做 deposit 動作的時候，我在程式碼中只有看到有紀錄和轉移 ETH(WETH) 到 AMM POOL 中。想請教一下若有人要對 AMM Pool 做 exchange (ETH -> crvUSD) 時。裡面的 crvUSD 是從何而來呢？ 謝謝。

想到可能的原因是一開始 AMM Pool 池裡面只有會 ETH (create_loan 產生）。此時這些 ETH 應該都會在 external price band 還要小的區間（離清算空間還很遠），所以都是 ETH 也是合理。需要到某天價格下跌到這些 ETH 的清算區間，套利者才會用他們的 crvUSD 來換取 ETH，此時 AMM Pool 中才開始會有 crvUSD。之後價格會回升套利者則會去把裡面的 crvUSD 給套利出來。這樣的機制下 AMM Pool 中的 crvUSD 都會是由套利者提供。
不知道我這樣的理解是正確的嗎？謝謝

**Reply by yurenju** [reply]
@ksin751119 你的理解是對的，這個 ETH 會被放在一個還沒有被觸及到的區間，比如說現在 ETH = 3000 crvUSD，存入之後 ETH 可能會被放在 500-550 crvUSD 的區間。只有到這個範圍的時候，ETH 才會用比市價（來自 Oracle 的價格資訊）更便宜的價格售出。

至於預言機的價格還沒落到這個區間前，我沒有仔細看這個 AMM 會如何反應，但猜測應該是會不能交易，或是可以交易取得的 ETH 完全不合理，讓使用者不會想要進來交易。

**Reply by ksin751119** [reply]
了解！感謝大大的說明！
順便想確認一下若 ETH = 3000，存進來後的 ETH 應該會放在小於 3000 的區間對嗎？因為如果放在大於 3000 的區間，照理來說應該都會是變成 crvUSD。

想請教一下另外兩個相關問題
1. 因為 LLAMMA pool price 通常是會比 oracle price 上升還要更高或下降得更低來製造套利空間讓套利者做 swap (部分清算）。因為 pool price 是由 oracle price 決定。若價格一下上升很多，池子裡面是不是有機會沒有足夠的 USD 來給套利者兌換 ETH？
舉例來說若 oracle price 等於 pool price 等於 1000。若 oracle price 下降到 800，此時 pool price 可能是 750。套利者用 750 crvUSD 在 pool  swap 1 ETH 出去。之後 oracle price 上升回 1000 ， pool price 假設變成 1100。此時 pool 中只有剛剛套利者丟進來的 750 crvUSD，不夠讓套利者以 1ETH 換到 1100 crvUSD（因為池子深度不夠）而造成高滑價。我的理解正確嗎？

2. 基本上 LLAMMA pool 就是一個高買低賣 ETH 的 Pool。 若區間一直不斷上升下降，是不是有可能造成在池子中的 ETH 會越來越少？承上例來說在 ETH = 1000 的時候，可能就無法換回 1 ETH（因為crvUSD 只有750個）

大大真不好意思... 最近剛好在理解 LLAMMA 的做法，有些地方還在試圖理解加上身邊沒有人可以一起討論，所以問題有點多。謝謝大大！

**Reply by yurenju** [reply]
> 了解！感謝大大的說明！ 順便想確認一下若 ETH = 3000，存進來後的 ETH 應該會放在小於 3000 的區間對嗎？因為如果放在大於 3000 的區間，照理來說應該都會是變成 crvUSD。

對的沒錯

> 想請教一下另外兩個相關問題
> 
> 1. 因為 LLAMMA pool price 通常是會比 oracle price 上升還要更高或下降得更低來製造套利空間讓套利者做 swap (部分清算）。因為 pool price 是由 oracle price 決定。若價格一下上升很多，池子裡面是不是有機會沒有足夠的 USD 來給套利者兌換 ETH？
>    舉例來說若 oracle price 等於 pool price 等於 1000。若 oracle price 下降到 800，此時 pool price 可能是 750。套利者用 750 crvUSD 在 pool  swap 1 ETH 出去。之後 oracle price 上升回 1000 ， pool price 假設變成 1100。此時 pool 中只有剛剛套利者丟進來的 750 crvUSD，不夠讓套利者以 1ETH 換到 1100 crvUSD（因為池子深度不夠）而造成高滑價。我的理解正確嗎？

在這個狀況針對 Pool 的行為我不確定他會怎麼作，可能需要進一步看源碼來確認他的反應。但是在急升急跌的狀況下，我滿確定池子肯定會有損失的。主要是因為價格跳太快的時候，套利者如果沒足夠快速的套利，價格升的太快會導致 LLAMM Pool 要用更好的價格收購，導致虧損。這個正好在 [PR review](https://github.com/lun-dao/LunDAO/pull/101#issuecomment-1337443232) 的時候 @paco0x 有提出解釋，你可以參考他舉的例子。

> 2. 基本上 LLAMMA pool 就是一個高買低賣 ETH 的 Pool。 若區間一直不斷上升下降，是不是有可能造成在池子中的 ETH 會越來越少？承上例來說在 ETH = 1000 的時候，可能就無法換回 1 ETH（因為crvUSD 只有750個）

理論上如果池子裡的 crvUSD 夠多的話應該是不會，正常的狀況池子裡應該會有許多的流動性才對。當然極端的狀況，比如說池子才剛開始成立時，或是遇到錯誤大家都把錢領走時確實會遇到。所以你說的是對的，但實際上應該只有極端狀況才會遇到。

> 大大真不好意思... 最近剛好在理解 LLAMMA 的做法，有些地方還在試圖理解加上身邊沒有人可以一起討論，所以問題有點多。謝謝大大！

不會，可以多多討論啊，我之前看的時候也沒什麼中文資料，感謝 paco 幫忙看文章才能找出許多盲點，有人可以一起討論會比較省力 😁
